### PR TITLE
Bucket name from secret

### LINKED ISSUE #1677: [Feature] Allow to specify bucket name in secret
In our setup, we will run argo in multitenant environment. That means, one set of users will have access to their namespace, but others won't. Argo secret management already requires S3 secrets to be present in namespace, allowing therefore to have different secret in namespace foo than bar (assuming secret is under same name and holds same keys).

We would like to use this to allow multitenancy in a way, that user group A will have their own dedicated bucket with IAM user tied to it. To achieve this, we would need ability to specify bucket name per namespace too, ideally in secret or configmap.

We could use configuration option like
```
bucketSecret:
   name: awsSecrets
   key: bucketName
```

That would work in similar way as credential secrets now, but would be alternative to `bucket` option.

### PR DESCRIPTION
**Pull request by inc0**
Fixes #1677

This change adds new configuration option `bucketSecret` which is an alternative to `bucket`.

It allows users to have different buckets per namespace, allowing for more multitenancy in workflow runs.

API changes:
Removed `required` from `bucket`
Added new option `bucketSecret` to provide secret key reference holding bucket name

### CODE REVIEW

File: workflow/artifacts/s3/s3.go [hunk 1]
```diff
@@ -38,13 +41,23 @@ func (s3Driver *S3ArtifactDriver) newS3Client() (argos3.S3Client, error) {
 func (s3Driver *S3ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
 	err := wait.ExponentialBackoff(wait.Backoff{Duration: time.Second * 2, Factor: 2.0, Steps: 5, Jitter: 0.1},
 		func() (bool, error) {
+			var err error
+			var bucket []byte
+			if inputArtifact.S3.BucketSecret.Key != "" {
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: *Question/Suggestion* As not one super savvy in Go I am always unclear if this is best practice or if there is a better way to determine if something does not exist (rather than it being an empty string).

File: workflow/artifacts/s3/s3.go [hunk 2]
```diff
@@ -38,13 +41,23 @@ func (s3Driver *S3ArtifactDriver) newS3Client() (argos3.S3Client, error) {
...
+				bucket = []byte(inputArtifact.S3.Bucket)
+			}
 			log.Infof("S3 Load path: %s, key: %s", path, inputArtifact.S3.Key)
 			s3cli, err := s3Driver.newS3Client()
 			if err != nil {
 				log.Warnf("Failed to create new S3 client: %v", err)
 				return false, nil
 			}
-			origErr := s3cli.GetFile(inputArtifact.S3.Bucket, inputArtifact.S3.Key, path)
+			origErr := s3cli.GetFile(string(bucket), inputArtifact.S3.Key, path)
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: *Question* Any reason not to convert `bucket` to a `string` above on lines 47 or 52?

File: workflow/artifacts/s3/s3.go [hunk 3]
```diff
@@ -53,17 +66,17 @@ func (s3Driver *S3ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string
 				return false, nil
 			}
 			// If we get here, the error was a NoSuchKey. The key might be a s3 "directory"
-			isDir, err := s3cli.IsDirectory(inputArtifact.S3.Bucket, inputArtifact.S3.Key)
+			isDir, err := s3cli.IsDirectory(string(bucket), inputArtifact.S3.Key)
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: Same as https://github.com/argoproj/argo/pull/1692/files#r337110522

File: workflow/artifacts/s3/s3.go [hunk 4]
```diff
@@ -53,17 +66,17 @@ func (s3Driver *S3ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string
...
+				log.Warnf("Failed to test if %s is a directory: %v", string(bucket), err)
 				return false, nil
 			}
 			if !isDir {
 				// It's neither a file, nor a directory. Return the original NoSuchKey error
 				return false, origErr
 			}
 
-			if err = s3cli.GetDirectory(inputArtifact.S3.Bucket, inputArtifact.S3.Key, path); err != nil {
+			if err = s3cli.GetDirectory(string(bucket), inputArtifact.S3.Key, path); err != nil {
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: Same as https://github.com/argoproj/argo/pull/1692/files#r337110522

File: workflow/artifacts/s3/s3.go [hunk 5]
```diff
@@ -77,6 +90,16 @@ func (s3Driver *S3ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string
 func (s3Driver *S3ArtifactDriver) Save(path string, outputArtifact *wfv1.Artifact) error {
 	err := wait.ExponentialBackoff(wait.Backoff{Duration: time.Second * 2, Factor: 2.0, Steps: 5, Jitter: 0.1},
 		func() (bool, error) {
+			var err error
+			var bucket []byte
+			if outputArtifact.S3.BucketSecret.Key != "" {
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: *Nit/Suggestion* Maybe we can make a helpful utility func to `getBucket` (as this logic is repeated above https://github.com/argoproj/argo/pull/1692/files#diff-0d7017944420a424dbd879fc0ef1c0ddR44)

File: workflow/artifacts/s3/s3.go [hunk 6]
```diff
@@ -89,12 +112,12 @@ func (s3Driver *S3ArtifactDriver) Save(path string, outputArtifact *wfv1.Artifac
 				return false, nil
 			}
 			if isDir {
-				if err = s3cli.PutDirectory(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
+				if err = s3cli.PutDirectory(string(bucket), outputArtifact.S3.Key, path); err != nil {
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: Same as https://github.com/argoproj/argo/pull/1692/files#r337110711

File: workflow/artifacts/s3/s3.go [hunk 7]
```diff
@@ -89,12 +112,12 @@ func (s3Driver *S3ArtifactDriver) Save(path string, outputArtifact *wfv1.Artifac
...
 					log.Warnf("Failed to put directory: %v", err)
 					return false, nil
 				}
 			} else {
-				if err = s3cli.PutFile(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
+				if err = s3cli.PutFile(string(bucket), outputArtifact.S3.Key, path); err != nil {
```
Comment (lorenanicole) [inline workflow/artifacts/s3/s3.go]: Same as https://github.com/argoproj/argo/pull/1692/files#r337111473

File: workflow/controller/operator_test.go [hunk 8]
```diff
@@ -1313,6 +1313,65 @@ func TestArtifactRepositoryRef(t *testing.T) {
...
+			},
+		},
+	)
+	woc.operate()
+	assert.Equal(t, woc.artifactRepository.S3.Endpoint, "my-minio-endpoint.default:9000")
+	assert.Equal(t, woc.artifactRepository.S3.BucketSecret.Key, "bucketname")
+	assert.Equal(t, wfv1.NodeRunning, woc.wf.Status.Phase)
+	pods, err := woc.controller.kubeclientset.CoreV1().Pods(wf.ObjectMeta.Namespace).List(metav1.ListOptions{})
+	assert.Nil(t, err)
+	assert.True(t, len(pods.Items) > 0, "pod was not created successfully")
```
Comment (lorenanicole) [inline workflow/controller/operator_test.go]: *Question* Will this always create just one pod? If so, can we tighten this to say `len(pods.Items) == 1` or whatever value it should be ... if it is determinate

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 9]
```diff
@@ -869,7 +872,7 @@ func (s *S3Artifact) String() string {
 }
 
 func (s *S3Artifact) HasLocation() bool {
-	return s != nil && s.Bucket != ""
+	return s != nil && (s.Bucket != "" || s.BucketSecret.Key != "")
```
Comment (juxtin) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: is it possible here that `s.BucketSecret` could be `nil`?

Comment (inc0) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I don't think so, swagger seems to be always initializing BucketSecret, but with empty variables

File: workflow/artifacts/s3/s3.go (hunk 1 above)
Comment (inc0) [reply workflow/artifacts/s3/s3.go]: hmm...I don't know, seems to be standard in codebase, but if someone knows better way, I'd love to know too

Comment (juxtin) [reply workflow/artifacts/s3/s3.go]: you can check `len(x) > 0`, which will return `false` if `x` is either `nil` or the empty string

File: workflow/controller/operator_test.go (hunk 8 above)
Comment (inc0) [reply workflow/controller/operator_test.go]: well, this is test case so if it creates a pod successfully, it's ok, at least for this test case

File: workflow/artifacts/s3/s3.go (hunk 5 above)
Comment (febuiles) [reply workflow/artifacts/s3/s3.go]: :+1: on @lorenanicole's comment, might be worth abstracting this to another function where we can do proper validation/creation/fetching.

File: workflow/controller/operator_test.go [hunk 10]
```diff
@@ -1407,6 +1407,65 @@ func TestArtifactRepositoryRef(t *testing.T) {
...
```
Comment (febuiles) [inline workflow/controller/operator_test.go]: (N) You can use `assert.Len` here.

**Comment by jessesuen** [comment]
Argo does not consider bucket names as sensitive. Even if theres a level of indirection from the spec to reference a bucket by a secret, it does not help because the workflow status fields store the bucket names for artifact locations. So the bucket names are exposed there. 

As for multi-tenancy issue, theres a (little known) feature which allows a configmap to be referenced in the workflow spec, that is supposed to solve this use case:

```yaml
spec:
  artifactRepositoryRef:
    configMap: my-confimap
    key: my-key
```

in which the key `my-key` in the ConfigMap `my-confimap` would contain artifact repository configuration.
```yaml
data:
  my-key: |
    artifactRepository:
      s3:
        bucket: my-bucket
...
```

`artifactRepositoryRef` exists because centralized controller configuration does not work for multi-tenant clusters.

**Comment by inc0** [comment]
Alternatively, if secrets are no-go, we could use `configmapKeyRef` in similar way - have configmap in each namespace with bucket value in it. Would that be better?

File: pkg/apis/workflow/v1alpha1/openapi_generated.go [hunk 11]
```diff
@@ -1459,7 +1465,7 @@ func schema_pkg_apis_workflow_v1alpha1_S3Artifact(ref common.ReferenceCallback)
 						},
 					},
 				},
-				Required: []string{"endpoint", "bucket", "accessKeySecret", "secretKeySecret", "key"},
+				Required: []string{"endpoint", "bucket", "bucketSecret", "accessKeySecret", "secretKeySecret", "key"},
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/openapi_generated.go]: why is this required here?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 12]
```diff
@@ -851,6 +851,9 @@ type S3Bucket struct {
 	// Bucket is the name of the bucket
 	Bucket string `json:"bucket"`
 
+	// Bucket secret reference
+	BucketSecret apiv1.SecretKeySelector `json:"bucketSecret"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: probably should have `json:"bucketSecret,emitempty"` - otherwise this is required, also bucket should have `omitempty`

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: probably should be `*apiv1.SecretKeySelector` than `apiv1.SecretKeySelector`

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 9 above)
Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I think that `bucketSecret` should be nillable.

Comment (inc0) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: if you make it nillable, checking accessing it's keys will always be risky of raising hard to debug exceptions. Way it's done now is that Key will default to "" if no bucketSecret is provided in config, so it's easier to handle. Also I remember having issues with swagger json def and nillable values. I couldn't define #/definitions/io.k8s.api.core.v1.SecretKeySelector or nil in swagger.json

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: @jessesuen suggested that it should not be nillable if there is a zero value, e.g. string=="" of bool=false, would I be right in thinking that's the case for bucketSecret?

Comment (inc0) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I don't understand the question. If you ask if there is default value that it will take if person doesn't specify `bucketSecret` in config at all, then answer is yes. If you don't have `bucketSecret` in config, Key will be "" and everything works as intended.

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Item 1.

I think my query comes from the swagger.json. If you can have either bucket or bucketKey, then they should not be required in the swagger.json. So i think they should have `omitempty` on them.

Item 2.

I wanted to check that it would be possible to omit them. I think you have confirm this.

**Comment by alexec** [comment]
@inc0 is this still wanted?

**Comment by inc0** [comment]
I'll be able to get back to it soon, but we'll need some way to provide namespaced bucket...

**Comment by alexec** [comment]
Adding `wontfix` as there has been no activity for 6 months.

**Comment by alexec** [comment]
No activity in 6 months - closing.
